File: src/runner.py
"""Task runner: launches Claude Code headless for a single task × profile combination.

The A-0 compiler didn't need to be elegant. It needed to work.
Each run gets its own directory so parallel runs don't step on each other.
"""

import subprocess
from dataclasses import dataclass, field
from pathlib import Path


@dataclass
class TaskResult:
    """Structured result from a single headless Claude Code execution."""

    task_id: str
    profile_name: str
    exit_code: int
    stdout: str
    stderr: str
    working_dir: str
    # Files written by Claude during the run
    files_written: list[str] = field(default_factory=list)


def run_task(
    task_id: str,
    task_spec: str,
    profile_name: str,
    system_prompt: str,
    working_dir: str,
    timeout: int = 300,
) -> TaskResult:
    """Run a single task with Claude Code headless.

    Launches `claude --print` with the task spec piped via stdin.
    If system_prompt is non-empty, passes it via --system-prompt.
    Captures stdout, stderr, exit code, and files written to the run directory.

    Each call creates an isolated subdirectory: working_dir/task_id/profile_name.

    Path anchoring (fix for #2): the task spec is prepended with an absolute
    WORKING DIRECTORY header before being piped to claude. The zero profile
    (empty system_prompt) already writes correctly to the subprocess cwd, but
    profiles with long system prompts were observed to sometimes write their
    solution/ directory at the repo root instead of the run_dir. Prepending
    an explicit absolute path and "write every file inside this directory"
    directive overrides whatever path resolution the agent was doing and
    anchors file operations to the intended run_dir.
    """
    run_dir = Path(working_dir) / task_id / profile_name
    run_dir.mkdir(parents=True, exist_ok=True)

    # Prepend explicit working-directory guidance. This is the preventive
    # fix for the stray-files-at-repo-root leak documented in issue #2.
    effective_spec = (
        f"WORKING DIRECTORY: {run_dir.absolute()}\n"
        f"All file paths referenced in the task below are RELATIVE to this "
        f"directory. Write every file you produce inside this directory. "
        f"Never write files outside this directory.\n"
        f"\n"
        f"---\n"
        f"\n"
        f"{task_spec}"
    )

    cmd = [
        "claude", "--print",
        "--dangerously-skip-permissions",
        "--no-session-persistence",
    ]
    if system_prompt:
        cmd += ["--system-prompt", system_prompt]

    result = subprocess.run(
        cmd,
        input=effective_spec,   # pipe spec via stdin — avoids CLI arg length limits
        cwd=str(run_dir),
        capture_output=True,
        text=True,
        timeout=timeout,
    )

    # Collect every file Claude wrote in the run directory.
    files_written = [
        str(f.relative_to(run_dir))
        for f in run_dir.rglob("*")
        if f.is_file()
    ]

    return TaskResult(
        task_id=task_id,
        profile_name=profile_name,
        exit_code=result.returncode,
        stdout=result.stdout,
        stderr=result.stderr,
        working_dir=str(run_dir),
        files_written=files_written,
    )

File: internal/runner/runner.go
package runner

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"syscall"
	"time"
)

const defaultTimeout = 300 * time.Second

// TaskResult holds the output from one headless Claude Code execution.
type TaskResult struct {
	TaskID       string
	ProfileName  string
	ExitCode     int
	Stdout       string
	Stderr       string
	WorkingDir   string
	FilesWritten []string
}

// RunTaskArgs groups the parameters for RunTask.
type RunTaskArgs struct {
	TaskID       string
	TaskSpec     string
	ProfileName  string
	SystemPrompt string
	WorkingDir   string
	Timeout      time.Duration // zero means use defaultTimeout
}

// RunTask runs one task+profile combination with Claude Code headless.
// Creates an isolated subdirectory: WorkingDir/TaskID/ProfileName.
func RunTask(ctx context.Context, args RunTaskArgs) (*TaskResult, error) {
	timeout := args.Timeout
	if timeout == 0 {
		timeout = defaultTimeout
	}

	runDir := filepath.Join(args.WorkingDir, args.TaskID, args.ProfileName)
	if err := os.MkdirAll(runDir, 0755); err != nil {
		return nil, fmt.Errorf("creating run dir: %w", err)
	}
	absRunDir, err := filepath.Abs(runDir)
	if err != nil {
		return nil, fmt.Errorf("resolving run dir: %w", err)
	}

	// Prepend working-directory anchor (fix for stray-files issue #2).
	effectiveSpec := fmt.Sprintf(
		"WORKING DIRECTORY: %s\n"+
			"All file paths referenced in the task below are RELATIVE to this "+
			"directory. Write every file you produce inside this directory. "+
			"Never write files outside this directory.\n"+
			"\n"+
			"---\n"+
			"\n"+
			"%s",
		absRunDir,
		args.TaskSpec,
	)

	cmdArgs := []string{
		"--print",
		"--dangerously-skip-permissions",
		"--no-session-persistence",
	}
	if args.SystemPrompt != "" {
		cmdArgs = append(cmdArgs, "--system-prompt", args.SystemPrompt)
	}

	runCtx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	cmd := exec.CommandContext(runCtx, "claude", cmdArgs...)
	cmd.Dir = absRunDir
	cmd.Stdin = strings.NewReader(effectiveSpec)
	// Put the child in its own process group so we can kill the whole
	// group (including grandchildren like a shell-spawned sleep) when
	// the context deadline fires.
	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
	// Override the default Cancel behaviour: send SIGKILL to the whole
	// process group rather than just the direct child.
	cmd.Cancel = func() error {
		if cmd.Process == nil {
			return nil
		}
		return syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
	}
	// WaitDelay forces the I/O pipes closed after the timeout so
	// cmd.Run() returns even if a grandchild is still holding them.
	cmd.WaitDelay = 2 * time.Second
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	runErr := cmd.Run()

	// Distinguish timeout from other errors.
	if runCtx.Err() == context.DeadlineExceeded {
		return &TaskResult{
			TaskID:       args.TaskID,
			ProfileName:  args.ProfileName,
			ExitCode:     124,
			Stdout:       "",
			Stderr:       fmt.Sprintf("TIMEOUT: claude process exceeded %s limit", timeout),
			WorkingDir:   absRunDir,
			FilesWritten: nil,
		}, nil
	}

	exitCode := 0
	if runErr != nil {
		if exitErr, ok := runErr.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			exitCode = 1
		}
	}

	// Collect files written under run dir.
	var filesWritten []string
	_ = filepath.Walk(absRunDir, func(path string, info os.FileInfo, err error) error {
		if err != nil || info.IsDir() {
			return nil
		}
		rel, _ := filepath.Rel(absRunDir, path)
		filesWritten = append(filesWritten, rel)
		return nil
	})

	return &TaskResult{
		TaskID:       args.TaskID,
		ProfileName:  args.ProfileName,
		ExitCode:     exitCode,
		Stdout:       stdout.String(),
		Stderr:       stderr.String(),
		WorkingDir:   absRunDir,
		FilesWritten: filesWritten,
	}, nil
}
